### PR TITLE
🛡️ Sentinel: [HIGH] Fix password leak in redact_url on parse failure

### PR DESCRIPTION
**Pull request by dhruvhaldar**
🚨 Severity: HIGH

💡 Vulnerability: `redact_url` in `ivi_water/security_utils.py` leaked the full URL, including credentials, when `urllib.parse.urlparse` raised an exception (e.g., due to an invalid port number) or failed to identify the password component (e.g., in schemeless URLs like `user:pass@host`). This violates the "Fail Securely" principle.

🎯 Impact: Sensitive credentials (passwords, tokens) embedded in malformed connection strings could be exposed in logs or error messages.

🔧 Fix:
-   Modified `redact_url` to return a safe placeholder `<REDACTION FAILED>` instead of the original URL upon exception.
-   Added fallback logic to prepend `//` to schemeless URLs containing `@`, forcing `urlparse` to recognize the `netloc` and identify credentials for redaction.

✅ Verification:
-   Added `tests/test_redact_url_vulnerability.py` covering invalid ports and schemeless URLs.
-   Ran existing test suite to ensure no regressions.

---
*PR created automatically by Jules for task [10943647803151677700](https://jules.google.com/task/10943647803151677700) started by @dhruvhaldar*

## Summary by CodeRabbit

* **Bug Fixes**
  * Prevents sensitive data leakage for malformed or edge-case URLs by returning a safe placeholder on parse failure.
  * Improves credential redaction for URLs lacking schemes or with unusual formatting, including cleanup of extraneous leading slashes after fallback parsing.

* **Tests**
  * Added comprehensive tests covering malformed URLs, invalid ports/IPs, control characters, and other edge cases to ensure redaction integrity.

<sub>✏️ Tip: You can customize this high-level summary in your review settings.</sub>

### CODE REVIEW

File: tests/test_redact_url_vulnerability.py [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    def test_redact_url_control_chars(self):
+        """Test that control characters don't cause issues."""
+        url = "https://user:secret_password@example.com/\x00"
+        redacted = redact_url(url)
+        assert "secret_password" not in redacted
+        assert "***REDACTED***" in redacted or "<REDACTION FAILED>" in redacted or redacted == url
+        # If it returns original url but password is NOT leaked (e.g. if password was part of leak), fail.
+        # But here password is clearly visible if returned as is.
+
```
[the remark this answers is not in the document]
Comment (dhruvhaldar) [reply tests/test_redact_url_vulnerability.py:39]: In `@tests/test_redact_url_vulnerability.py` around lines 33 - 41, The test
currently allows returning the original URL (via `or redacted == url`), which
permits password leakage; update the `test_redact_url_control_chars` assertions
to require that the returned value does not equal the original and does not
contain the secret (remove `redacted == url`), and instead assert that the
redaction produced an expected redaction marker (e.g., `"***REDACTED***"` or
`"<REDACTION FAILED>"`) or a safely transformed URL; ensure this targets the
`redact_url` behavior so the test fails if the password remains in the output.
